main.go

```diff
@@ -1,115 +1,144 @@
-	heartbeatProcessor := processors.NewHeartbeatProcessor()
-	counterProcessor := processors.NewCounterProcessor()
 
 	sender := statsd.NewStatsdClient(*statsdEndpoint, *statsdPrefix)
 	sender.CreateSocket()
 
 	var processedMetrics []metrics.Metric
 	var proc_err error
 
-	msgChan := make(chan *events.Envelope)
-	go func() {
-		defer close(msgChan)
-		errorChan := make(chan error)
-		go consumer.Firehose(*subscriptionId, authToken, msgChan, errorChan, nil)
+	msgChan := make(chan *metrics.Stream)
+	errorChan := make(chan error)
 
+	go func() {
 		for err := range errorChan {
 			fmt.Fprintf(os.Stderr, "%v\n", err.Error())
 		}
 	}()
 
-	for msg := range msgChan {
-		eventType := msg.GetEventType()
-
-		// graphite-nozzle can handle CounterEvent, ContainerMetric, Heartbeat,
-		// HttpStartStop and ValueMetric events
-		switch eventType {
-		case events.Envelope_ContainerMetric:
-			processedMetrics, proc_err = containerMetricProcessor.Process(msg)
-		case events.Envelope_CounterEvent:
-			processedMetrics, proc_err = counterProcessor.Process(msg)
-		case events.Envelope_Heartbeat:
-			processedMetrics, proc_err = heartbeatProcessor.Process(msg)
-		case events.Envelope_HttpStartStop:
-			processedMetrics, proc_err = httpStartStopProcessor.Process(msg)
-		case events.Envelope_ValueMetric:
-			processedMetrics, proc_err = valueMetricProcessor.Process(msg)
-		default:
-			// do nothing
+	go func() {
+		apps := make(map[string]*consumer.Consumer)
+
+		for {
+			err := updateApps(client, apps, msgChan, errorChan)
+			if err != nil {
+				fmt.Println(err)
+				os.Exit(-1)
+			}
+
+			time.Sleep(time.Duration(*updateFrequency) * time.Second)
 		}
+	}()
+
+	for wrapper := range msgChan {
+		processedMetrics, proc_err = containerMetricProcessor.Process(wrapper)
 
 		if proc_err != nil {
 			fmt.Fprintf(os.Stderr, "%v\n", proc_err.Error())
 		} else {
-			if !*debug {
-				if len(processedMetrics) > 0 {
-					for _, metric := range processedMetrics {
-						var prefix string
-						if *prefixJob {
-							prefix = msg.GetJob() + "." + msg.GetIndex()
+			if len(processedMetrics) > 0 {
+				for _, metric := range processedMetrics {
+					if *debug {
+						fmt.Println(metric)
+					} else {
+						err := metric.Send(sender)
+						if err != nil {
+							fmt.Fprintf(os.Stderr, "%v\n", err)
 						}
-						metric.Send(sender, prefix)
 					}
 				}
-			} else {
-				for _, msg := range processedMetrics {
-					fmt.Println(msg)
-				}
 			}
 		}
 		processedMetrics = nil
 	}
 }
+
+func updateApps(client *cfclient.Client, watchedApps map[string]*consumer.Consumer, msgChan chan *metrics.Stream, errorChan chan error) error {
+	authToken, err := client.GetToken()
+	if err != nil {
+		return err
+	}
+
+	apps, err := client.ListApps()
+	if err != nil {
+		return err
+	}
+
+	runningApps := map[string]bool{}
+	for _, app := range apps {
+		runningApps[app.Guid] = true
+		if _, ok := watchedApps[app.Guid]; !ok {
```

Comment:
You might be right. Although, from the top of my head, this may not be as simple as that.

We do have our loop checking for errors: [here](https://github.com/alphagov/paas-cf-apps-statsd/blob/metric_exporter_139278341/main.go#L56-L60) at the current state of the NOAA library, we won't really know which application returned that error...

Is that something you'd like to see in this PR or could it be possibly scheduled into a follow-up story?